packages/flutter_tools/lib/src/android/deferred_components.dart

```diff
@@ -0,0 +1,674 @@
+  androidStringMapping,
+  loadingUnitGolden,
+  writeGolden,
+}
+
+/// A class to configure and run a deferred component setup verification check.
+///
+/// To configure a setup verification run, supply a list of DeferredComponentsSetupVerificationCheck
+/// values to the contstructor. To run the the setup verification, invoke [runSetup] which
+/// will run each check in the order provided. The results of the checks will be displayed.
+///
+/// Failing a check will result in the tool to exit, ending the current command unless false is
+/// passed to the [exitOnFail] parameter.
+class DeferredComponentsSetupVerification {
+  DeferredComponentsSetupVerification(this.checks, this.env, {
+    this.title = 'Deferred components setup verification',
+  });
+
+  final List<DeferredComponentsSetupVerificationCheck> checks;
+
+  final Environment env;
+
+  final String title;
+
+  /// The name of the golden file that tracks the latest loading units generated.
+  String _kDeferredComponentsGoldenFileName = 'deferred_components_golden.yaml';
+  /// The directory in the build folder to generate missing/modified files into.
+  String _kDeferredComponentsTempDirectory = 'android_deferred_components_setup_files';
+
+  Directory get _outputDir => env.projectDir
+      .childDirectory('build')
+      .childDirectory(_kDeferredComponentsTempDirectory);
+
+  void runSetup({List<DeferredComponent> components, List<LoadingUnit> generatedLoadingUnits, bool exitOnFail = true}) async {
+    List<String> generatedFiles = <String>[];
+    List<String> modifiedFiles = <String>[];
+    Map<String, dynamic> goldenComparisonResults;
+    for (DeferredComponentsSetupVerificationCheck check in checks) {
+      switch (check) {
+        case DeferredComponentsSetupVerificationCheck.androidRootSetup:
+          modifiedFiles.addAll(_checkAndroidRootSettings(components));
+          break;
+        case DeferredComponentsSetupVerificationCheck.androidComponentSetup:
+          generatedFiles = await _checkAndroidComponentSetup(components);
+          break;
+        case DeferredComponentsSetupVerificationCheck.androidStringMapping:
+          modifiedFiles.addAll(_checkAndroidStringMapping(components, generatedLoadingUnits));
+          break;
+        case DeferredComponentsSetupVerificationCheck.loadingUnitGolden:
+          Map<String, dynamic> deferredComponentsGolden = _parseGolden(env.projectDir.childFile(_kDeferredComponentsGoldenFileName));
+          goldenComparisonResults = _checkLoadingUnitGolden(generatedLoadingUnits, deferredComponentsGolden);
+          break;
+        case DeferredComponentsSetupVerificationCheck.writeGolden:
+          _writeGolden(generatedLoadingUnits);
+          break;
+      }
+    }
+    _handleResults(generatedFiles, modifiedFiles, goldenComparisonResults, exitOnFail);
+  }
+
+  // Generates additions to settings.gradle, app/build.gradle
+  List<String> _checkAndroidRootSettings(List<DeferredComponent> components) {
```

Comment:
Being able to visualize these steps is great!

I'd document all these steps to say what they do in their declarations or in the switch.